### ISSUE TITLE
Unable to retrieve assembly location of the symbol under analysis

### ISSUE THREAD

**Issue by msbasanth** [dotnet/roslyn#23741]
**Version Used**: 
Roslyn v2.6.0
netstandard1.6

**Steps to Reproduce**:

1. I have an analyzer which analyzes type symbol INamedType which in a different assembly.
2. Trying to get Assembly.Location of the symbol.ContainingAssembly as below,
System.Reflection.Assembly.Load(new AssemblyName(symbol.ContainingAssembly.MetadataName)); OR
Assembly.Load(new AssemblyName(symbol.ContainingAssembly.Identity.GetDisplayName()));
 * Both option for assembly load works fine if we are debugging in IDE (I could see the current directory is project folder that time.)
* But when I am not debugging Assembly load fails as the current working directory is either Roslyn bin folder or devenv.exe folder.
3. What is the good approach to retrieve Assembly.Location of the containing assembly of a symbol.

**Expected Behavior**:
Able to access assembly location through ISymbol.ContainingAssembly. 

**Actual Behavior**:
Assembly location is not accessible through ContainingAssembly. 

How can I access assembly location of the symbol under analysis?

Thanks
Basanth

**Comment by jcouv** [comment]
@mavasani Any advice?

**Comment by jmarolf** [comment]
`ISymbol.ContainingAssembly` returns an `IAssemblySymbol` which can give you the name but not the path.  Can you elaborate on why you need to load an dll to do analysis?

**Comment by msbasanth** [comment]
I am trying to retrieve XML Documentation of the symbol which is from a different assembly. I am not able to retrieve XMLDocumentation from symbol as XMLDocumentation provider is not configured.
The issue is https://github.com/dotnet/roslyn/issues/23673

So if we are able to get the symbol assembly path I could use XmlDocumentationProvider (with minor modification to return XML Documentation) for retrieving the XML comments.

string xmlDocumentationPath = Path.Combine(**binPath**, $"{symbol.ContainingAssembly.Name}.xml");
var xmlProvider = XmlDocumentationProvider.CreateFromFile(xmlDocumentationPath);
var xmlDocumentation = xmlProvider.GetDoc(symbol.GetDocumentationCommentId()); // I copied and extended XmlDocumentationProvider with GetDoc.

For this, I need binPath which is the assembly location.

**Comment by jmarolf** [comment]
@msbasanth in general analyzers look at source code (at least that is how the API is designed).  XML doc comment loading is something that is normally done by the IDE.  Is the hope to provide a fix for doc comments or show some information to the user?

**Comment by msbasanth** [comment]
Ya, @jmarolf we have an analyzer which needs to analyze XML comment for particular tags and show information. This works fine if the symbol is in the same solution and we can retrieve XMLDocumentationTrivia for analysis. But when the assembly is external we need to relay on XML documentation. i.e. <AssemblyName>.xml.

I think XmlDocumentationProvider is introduced for IDE to support such analysis. But if we can inject XmlDocumentationProvider somehow to the compilation we can get XML documentation using ISymbol.GetDocumentationCommentXml().

**Comment by jasonmalinowski** [comment]
Yeah, the right fix for this is for us to move the documentation provider down into the compiler. I can't recall if we layered it this way because we generally figured the compiler didn't need to be doing any doc comment file reading (which your scenario points out might have been the wrong decision), or if at the time the compiler was "too portable" and wasn't allowed to have file access like that.

@jaredpar thoughts here?

**Comment by jaredpar** [comment]
This issue seems to have the assumption that all assemblies have paths. That is not necessarily true. It's possible to load assembly from raw bytes in which case there is no path to resolve. That is a large reason why the interface doesn't have a path element, just a name. 

Moving the documentation provider down into the compiler won't change that.

**Comment by jasonmalinowski** [comment]
@jaredpar I was imagining moving the provider to the command line compiler and passing it to compilations made there. There we do have paths, I'd hope?

**Comment by msbasanth** [comment]
We have a good number of FxCop rules which analyzes XML documentation which we want to port to analyzers. The analyzers which analyze the source symbols we are able to retrieve the XML documentation through XMLDocumentationTrivia. When we analyze the metadata where the XML documentation is part of <AssemblyName>.xml file we struggle.

I personally feel XmlDocumentationProvider is very much valid for custom analyzers, with which you can provide XML documentation even for symbols from a custom assembly referenced. i.e. by making API ISymbol.GetDocumentationComment() valid even for custom assemblies which provides XML Documentation externally.

**Comment by msbasanth** [comment]
I could get the assembly documentation XML path by,

1. Read (semanticModel.Compilation.Options.XmlReferenceResolver as XmlFileReader).BaseDirectory which gives the project directory.
2. Used Platform(x86) & OptimizationLevel (e.g. Debug) from semanticModel.Compilation.Options to retrieve the correct output path by parsing the project file.
3. Output folder + $"{symbol.ContainingAssembly.Name}.xml" will give you the XMLDocumentation path.

Can I rely on XmlReferenceResolver.BaseDirectory for fetching the project folder?

**Comment by msbasanth** [comment]
Adding one observation here, @mavasani  @jaredpar 
As I mentioned above I am using rip of XmlDocumentationProvider for reading XML Documentation. 

Observation in XmllDocumentationProvider.cs with GetDocumentationForSymbol,

protected override string GetDocumentationForSymbol(string documentationMemberID, CultureInfo preferredCulture, CancellationToken cancellationToken = default)
        {
            if (_docComments == null)
            {
                using (_gate.DisposableWait(cancellationToken))
                { 
                   // We do the real read here which takes time.
                 }
             }
         }

With this I observed when multiple threads are accessing GetDocumentationForSymbol _docComments may have a partial data as the NonReEntrant check is happening after if (_docComments == null).

This really fixes this issue for me as the reentrant threads will wait untill all docComments filled then go for the check. 
using (_gate.DisposableWait(cancellationToken))
 {
     if (_docComments == null)
            {}
 }

Please redirect to right people if this observation points to a probable correction.

Regards
Basanth

**Comment by CyrusNajmabadi** [comment]
I don't believe we would do this based on only getting a single request on this.
